File: Linkedlist/83.py
"""
83. 删除排序链表中的重复元素
简单
https://leetcode-cn.com/problems/remove-duplicates-from-sorted-list/
"""

# Definition for singly-linked list.
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution:
    # 68 ms	  14.03
    # 13.8 MB 6.06
    def deleteDuplicates(self, head: ListNode) -> ListNode:
        last = head
        while last and last.next:
            if last.val == last.next.val:
                last.next = last.next.next
            else:
                last = last.next
        return head

File: Leetcode-Swift/Leetcode/0001-0100/83.swift
//
//  83.swift
//  Leetcode-Swift
//
//  Created by 95cc on 2021/12/7.
//

import Foundation

/*
 83. 删除排序链表中的重复元素 (简单)
 https://leetcode-cn.com/problems/remove-duplicates-from-sorted-list/
 */

class Solution_83 {
    func deleteDuplicates(_ head: ListNode?) -> ListNode? {
        // 参数校验
        if head == nil {
            return nil
        }
        
        var cur = head
        while cur != nil {
            while let next = cur?.next, cur!.val == next.val {
                cur?.next = cur?.next?.next
            }
            cur = cur?.next
        }
        
        return head
    }
    
    func test() {
        printListNode(rootNode: deleteDuplicates(createListNode(values: [1,1,2,3,3])))
        printListNode(rootNode: deleteDuplicates(createListNode(values: [1,1,2])))
        printListNode(rootNode: deleteDuplicates(createListNode(values: [1,1])))
    }
}

